{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2007-2008 Minor antigenic variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The goal of this work is to determine which HA variants are nonsymonous and whether or any of the nonsymonous variants are in putative antigenic regions. I am only looking at minor variants in this analyis and so nonsymonous variants will be identified as variants that change the amino acid a given position compared to the sample's consensus seqeunce (not the plasmid control)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Based on what I have seen in the Fluvac data the we have 1700bp of the HA sequences. All seqeunces start with \"ATT\" When we blast these sequences in NCBI. This first \"codon\" algins with the start of the coding sequence \"ATG\". If I make this change in the plasmid control .fa and then translate the altered string, the resulting protein sequence is an exact match to what is observed in NCBI. It looks like Kari aligned to just the coding region during the primary analysis.\n",
    "\n",
    "The first step will be to read in the parsed HA consensus sequences. They are in two fasta files named after the run from which they come (1293 & 1304). Here I am reading each sequence, adjusting the start codon, translating, and then saving the sequence names, nucleotide seq object, and protein seq object in separate lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/Bio/Seq.py:2071: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "Ids=[]\n",
    "nucleotide=[]\n",
    "protein=[]\n",
    "\n",
    "for seq_record in SeqIO.parse(\"/Users/jt/Documents/Analysis/FluVac_paper/data/processed/Run_1293/parsed_fa/HA.fa\", \"fasta\"):\n",
    "    Ids.append(seq_record.id.split(\".\")[0])\n",
    "    seq_record.seq.alphabet=IUPAC.unambiguous_dna\n",
    "    fixed = list(str(seq_record.seq)) # the sequences all start as ATT they should be ATG - this is likely due to issues of coverage near the end. Fixing this error opens up the proper reading frame and when done to the plasmid control it yeilds a 100% match across all 566aa found in NCBI\n",
    "    fixed[2]=\"G\"\n",
    "    fixed=''.join(fixed)\n",
    "    fixed_record=Seq(fixed,IUPAC.unambiguous_dna)\n",
    "    nucleotide.append(fixed_record)\n",
    "    prot=fixed_record.translate()\n",
    "    protein.append(prot)\n",
    "for seq_record in SeqIO.parse(\"/Users/jt/Documents/Analysis/FluVac_paper/data/processed/Run_1304/parsed_fa/HA.fa\", \"fasta\"):\n",
    "    Ids.append(seq_record.id.split(\".\")[0])\n",
    "    seq_record.seq.alphabet=IUPAC.unambiguous_dna\n",
    "    fixed = list(str(seq_record.seq)) # the sequences all start as ATT they should be ATG - this is likely due to issues of coverage near the end. Fixing this error opens up the proper reading frame and when done to the plasmid control it yeilds a 100% match across all 566aa found in NCBI\n",
    "    fixed[2]=\"G\"\n",
    "    fixed=''.join(fixed)\n",
    "    fixed_record=Seq(fixed,IUPAC.unambiguous_dna)\n",
    "    nucleotide.append(fixed_record)\n",
    "    prot=fixed_record.translate()\n",
    "    protein.append(prot)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I will read in all polymorphisms on HA from the 2007-2008 season."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "data=pd.read_csv(\"../data/2007_2008.ha.csv\")\n",
    "df=data.sort_values(\"Id\") # Sorting the data frame by sample name. NOTE : The indeces don't change \n",
    "df[\"pos\"]=df[\"pos\"]-1 # Note the pandas data frame is in base 1 while everthing else is in base 0 this fixes that\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Now to find the codon of each of these positions we will just need to use a little modular algebra. bp 0-2 are in represent the o aa 3-5 the 1 and so forth. So bp divided by 3 no remander should do the trick. This will give the amino acid position or \"aa_pos\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "df[\"aa_pos\"]=df[\"pos\"]//3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I am looping through the sample names (Id) in the variant csv. For each Id, I am finding the index of that Id in the Id list; This is also the index of the sample in the protein list. There are mulitple variants in some samples - by looping through the Ids in the data frame I will grab each of these instances separately. I am using a counter to keep track of what row of the data frame. I am then taking the appropriate amino acid position and grabing the consensus amino acid for that position in the sample. These amino acids are stored in a list and then added to the data frame in the last line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "counter=0\n",
    "consens=[]\n",
    "for id in df.Id:\n",
    "    if str(id) in Ids:\n",
    "        i=Ids.index(str(id)) # get the index of the sample\n",
    "        aa_pos=list(df[\"aa_pos\"])[counter]\n",
    "        aa=str(protein[i])[aa_pos]\n",
    "        #print(str(i)+ \" : \" + str(id) + \" : \"+str(aa_pos) + \" : \" + aa)\n",
    "        consens.append(aa)\n",
    "    else:\n",
    "        consens.append(\"NA\")\n",
    "    counter=counter+1\n",
    "df[\"aa_consensus\"]=consens"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I am looping through the variants by sample name (as above). This time I am introducing the mutation into the nucleaic acid consensus. I am then retranslating the seqeunce, and grabing the appropriate amino acid. As before each step is saved as a list and in the end the variant amino acid is added to the data frame. If the variant amino acid is different than the consensus, the sample Id, nucleaic acid, and amino acid sequences are also saved in separate lists to be used later (denoted as dn\\_*)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/Bio/Seq.py:2071: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning)\n"
     ]
    }
   ],
   "source": [
    "counter=0\n",
    "var_nucleotide=[]\n",
    "var_Ids=[]\n",
    "var_protein=[]\n",
    "var_aa=[]\n",
    "\n",
    "dn_nucleotide=[]\n",
    "dn_Ids=[]\n",
    "dn_protein=[]\n",
    "\n",
    "\n",
    "for id in df.Id:\n",
    "    if str(id) in Ids:\n",
    "        i=Ids.index(str(id)) # get the index of the sample\n",
    "        pos=list(df[\"pos\"])[counter]\n",
    "        var=list(df[\"var\"])[counter]\n",
    "        seq=list(str(nucleotide[i])) # get the consensus and make it a list so we can make the changes\n",
    "        seq[pos]=var\n",
    "        seq=''.join(seq)\n",
    "        seq=Seq(seq,IUPAC.unambiguous_dna)\n",
    "        var_nucleotide.append(seq)\n",
    "        prot=seq.translate()\n",
    "        aa_pos=list(df[\"aa_pos\"])\n",
    "        aa_pos=aa_pos[counter]\n",
    "        aa=str(prot)[aa_pos]\n",
    "        con_aa=list(df[\"aa_consensus\"])[counter]\n",
    "        if aa !=con_aa:\n",
    "            dn_protein.append(prot)\n",
    "            dn_Ids.append(str(id))\n",
    "        \n",
    "        #print(str(id) + \" : \"+str(aa_pos) + \" : \" + aa + \" : \" + str(counter))\n",
    "        var_aa.append(aa)\n",
    "        var_protein.append(prot)\n",
    "        var_Ids.append(str(id))\n",
    "    else:\n",
    "        var_nucleotide.append(\"NA\")\n",
    "        var_Ids.append(\"NA\")\n",
    "        var_protein.append(\"NA\")\n",
    "        var_aa.append(\"NA\")\n",
    "    counter=counter+1\n",
    "df[\"aa_var\"]=var_aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The HA variants "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Id is the sample name, ref is the reference base in the plasmid control, var is the variant in the sample, freq.var the frequency, pos the nucleaic acid postion, aa_pos the amino acid position, aa_consensus the amino acid at the consensus level, and aa_var the variant amino acid."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>ref</th>\n",
       "      <th>var</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>pos</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>aa_consensus</th>\n",
       "      <th>aa_var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.020529</td>\n",
       "      <td>1114</td>\n",
       "      <td>371</td>\n",
       "      <td>Q</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.985102</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.014898</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.979471</td>\n",
       "      <td>1114</td>\n",
       "      <td>371</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.039535</td>\n",
       "      <td>284</td>\n",
       "      <td>94</td>\n",
       "      <td>F</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.960465</td>\n",
       "      <td>284</td>\n",
       "      <td>94</td>\n",
       "      <td>F</td>\n",
       "      <td>F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.980604</td>\n",
       "      <td>1014</td>\n",
       "      <td>338</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>10</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.033524</td>\n",
       "      <td>731</td>\n",
       "      <td>243</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>10</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.966476</td>\n",
       "      <td>731</td>\n",
       "      <td>243</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.034423</td>\n",
       "      <td>687</td>\n",
       "      <td>229</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.965577</td>\n",
       "      <td>687</td>\n",
       "      <td>229</td>\n",
       "      <td>I</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.019396</td>\n",
       "      <td>1014</td>\n",
       "      <td>338</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>12</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.851825</td>\n",
       "      <td>1295</td>\n",
       "      <td>431</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>12</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.146007</td>\n",
       "      <td>1295</td>\n",
       "      <td>431</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.017493</td>\n",
       "      <td>1047</td>\n",
       "      <td>349</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.982507</td>\n",
       "      <td>1047</td>\n",
       "      <td>349</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.078873</td>\n",
       "      <td>853</td>\n",
       "      <td>284</td>\n",
       "      <td>R</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.921127</td>\n",
       "      <td>853</td>\n",
       "      <td>284</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.044289</td>\n",
       "      <td>1403</td>\n",
       "      <td>467</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.955711</td>\n",
       "      <td>1403</td>\n",
       "      <td>467</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.136396</td>\n",
       "      <td>1235</td>\n",
       "      <td>411</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.422878</td>\n",
       "      <td>1613</td>\n",
       "      <td>537</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.988190</td>\n",
       "      <td>1577</td>\n",
       "      <td>525</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.863604</td>\n",
       "      <td>1235</td>\n",
       "      <td>411</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.577122</td>\n",
       "      <td>1613</td>\n",
       "      <td>537</td>\n",
       "      <td>M</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.021698</td>\n",
       "      <td>1196</td>\n",
       "      <td>398</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>18</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.978302</td>\n",
       "      <td>1196</td>\n",
       "      <td>398</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>19</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.016013</td>\n",
       "      <td>1196</td>\n",
       "      <td>398</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>19</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.983987</td>\n",
       "      <td>1196</td>\n",
       "      <td>398</td>\n",
       "      <td>R</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>22</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.968217</td>\n",
       "      <td>1565</td>\n",
       "      <td>521</td>\n",
       "      <td>E</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>46</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.969850</td>\n",
       "      <td>879</td>\n",
       "      <td>293</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>49</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.011324</td>\n",
       "      <td>761</td>\n",
       "      <td>253</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>50</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.013804</td>\n",
       "      <td>728</td>\n",
       "      <td>242</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>50</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.986196</td>\n",
       "      <td>728</td>\n",
       "      <td>242</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>50</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.971189</td>\n",
       "      <td>761</td>\n",
       "      <td>253</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>50</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.028811</td>\n",
       "      <td>761</td>\n",
       "      <td>253</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>54</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.016006</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>54</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.983994</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>55</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.934185</td>\n",
       "      <td>410</td>\n",
       "      <td>136</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>55</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.065815</td>\n",
       "      <td>410</td>\n",
       "      <td>136</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.660234</td>\n",
       "      <td>1268</td>\n",
       "      <td>422</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.339766</td>\n",
       "      <td>1268</td>\n",
       "      <td>422</td>\n",
       "      <td>Q</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>66</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.010083</td>\n",
       "      <td>1228</td>\n",
       "      <td>409</td>\n",
       "      <td>Q</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>66</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.647101</td>\n",
       "      <td>1559</td>\n",
       "      <td>519</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>66</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.347261</td>\n",
       "      <td>1559</td>\n",
       "      <td>519</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>67</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.939217</td>\n",
       "      <td>1457</td>\n",
       "      <td>485</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>67</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.060783</td>\n",
       "      <td>1457</td>\n",
       "      <td>485</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>76</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.093062</td>\n",
       "      <td>1502</td>\n",
       "      <td>500</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>76</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.904180</td>\n",
       "      <td>1502</td>\n",
       "      <td>500</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>78</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.983737</td>\n",
       "      <td>1539</td>\n",
       "      <td>513</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>78</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.016263</td>\n",
       "      <td>1539</td>\n",
       "      <td>513</td>\n",
       "      <td>N</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>82</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.989965</td>\n",
       "      <td>1512</td>\n",
       "      <td>504</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>88</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.013609</td>\n",
       "      <td>864</td>\n",
       "      <td>288</td>\n",
       "      <td>P</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>88</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.986391</td>\n",
       "      <td>864</td>\n",
       "      <td>288</td>\n",
       "      <td>P</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>89</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>0.989561</td>\n",
       "      <td>869</td>\n",
       "      <td>289</td>\n",
       "      <td>I</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>89</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.972221</td>\n",
       "      <td>1335</td>\n",
       "      <td>445</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>89</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.027779</td>\n",
       "      <td>1335</td>\n",
       "      <td>445</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>89</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.010439</td>\n",
       "      <td>869</td>\n",
       "      <td>289</td>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>103</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "      <td>0.943433</td>\n",
       "      <td>1120</td>\n",
       "      <td>373</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>103</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.056567</td>\n",
       "      <td>1120</td>\n",
       "      <td>373</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>70 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Id ref var  freq.var   pos  aa_pos aa_consensus aa_var\n",
       "34    3   A   T  0.020529  1114     371            Q      L\n",
       "6     3   G   A  0.985102   670     223            K      K\n",
       "7     3   G   G  0.014898   670     223            K      R\n",
       "33    3   A   A  0.979471  1114     371            Q      Q\n",
       "2     6   C   A  0.039535   284      94            F      L\n",
       "3     6   C   C  0.960465   284      94            F      F\n",
       "30   10   G   G  0.980604  1014     338            V      V\n",
       "17   10   C   T  0.033524   731     243            S      S\n",
       "16   10   C   C  0.966476   731     243            S      S\n",
       "13   10   A   G  0.034423   687     229            I      V\n",
       "12   10   A   A  0.965577   687     229            I      I\n",
       "29   10   G   A  0.019396  1014     338            V      I\n",
       "48   12   C   A  0.851825  1295     431            T      T\n",
       "49   12   C   T  0.146007  1295     431            T      T\n",
       "32   12   G   T  0.017493  1047     349            A      S\n",
       "31   12   G   G  0.982507  1047     349            A      A\n",
       "21   14   G   A  0.078873   853     284            R      K\n",
       "22   14   G   G  0.921127   853     284            R      R\n",
       "52   16   G   A  0.044289  1403     467            K      K\n",
       "53   16   G   G  0.955711  1403     467            K      K\n",
       "45   17   A   G  0.136396  1235     411            E      E\n",
       "68   17   A   A  0.422878  1613     537            M      I\n",
       "67   17   A   G  0.988190  1577     525            G      G\n",
       "44   17   A   A  0.863604  1235     411            E      E\n",
       "69   17   A   G  0.577122  1613     537            M      M\n",
       "40   18   G   G  0.021698  1196     398            R      R\n",
       "39   18   G   A  0.978302  1196     398            R      R\n",
       "42   19   G   G  0.016013  1196     398            R      R\n",
       "41   19   G   A  0.983987  1196     398            R      R\n",
       "66   22   G   G  0.968217  1565     521            E      E\n",
       "..  ...  ..  ..       ...   ...     ...          ...    ...\n",
       "27   46   A   A  0.969850   879     293            N      N\n",
       "18   49   A   G  0.011324   761     253            K      K\n",
       "14   50   C   C  0.013804   728     242            P      P\n",
       "15   50   C   T  0.986196   728     242            P      P\n",
       "19   50   A   A  0.971189   761     253            K      K\n",
       "20   50   A   C  0.028811   761     253            K      N\n",
       "9    54   G   G  0.016006   670     223            K      R\n",
       "8    54   G   A  0.983994   670     223            K      K\n",
       "4    55   C   C  0.934185   410     136            N      N\n",
       "5    55   C   T  0.065815   410     136            N      N\n",
       "47   64   G   G  0.660234  1268     422            Q      Q\n",
       "46   64   G   A  0.339766  1268     422            Q      Q\n",
       "43   66   A   T  0.010083  1228     409            Q      L\n",
       "63   66   C   A  0.647101  1559     519            G      G\n",
       "64   66   C   T  0.347261  1559     519            G      G\n",
       "54   67   C   C  0.939217  1457     485            Y      Y\n",
       "55   67   C   T  0.060783  1457     485            Y      Y\n",
       "58   76   T   A  0.093062  1502     500            T      T\n",
       "59   76   T   C  0.904180  1502     500            T      T\n",
       "61   78   A   A  0.983737  1539     513            N      N\n",
       "62   78   A   C  0.016263  1539     513            N      H\n",
       "60   82   G   A  0.989965  1512     504            N      N\n",
       "24   88   C   T  0.013609   864     288            P      S\n",
       "23   88   C   C  0.986391   864     288            P      P\n",
       "26   89   T   T  0.989561   869     289            I      I\n",
       "51   89   G   G  0.972221  1335     445            A      A\n",
       "50   89   G   A  0.027779  1335     445            A      T\n",
       "25   89   T   G  0.010439   869     289            I      M\n",
       "36  103   C   C  0.943433  1120     373            S      S\n",
       "35  103   C   A  0.056567  1120     373            S      Y\n",
       "\n",
       "[70 rows x 8 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[\"Id\",\"ref\",'var',\"freq.var\",\"pos\",\"aa_pos\",\"aa_consensus\",\"aa_var\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>ref</th>\n",
       "      <th>var</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>pos</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>aa_consensus</th>\n",
       "      <th>aa_var</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.020529</td>\n",
       "      <td>1114</td>\n",
       "      <td>371</td>\n",
       "      <td>Q</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.014898</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.039535</td>\n",
       "      <td>284</td>\n",
       "      <td>94</td>\n",
       "      <td>F</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.034423</td>\n",
       "      <td>687</td>\n",
       "      <td>229</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.019396</td>\n",
       "      <td>1014</td>\n",
       "      <td>338</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.017493</td>\n",
       "      <td>1047</td>\n",
       "      <td>349</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.078873</td>\n",
       "      <td>853</td>\n",
       "      <td>284</td>\n",
       "      <td>R</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.422878</td>\n",
       "      <td>1613</td>\n",
       "      <td>537</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.030421</td>\n",
       "      <td>192</td>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>46</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.030150</td>\n",
       "      <td>879</td>\n",
       "      <td>293</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>50</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.028811</td>\n",
       "      <td>761</td>\n",
       "      <td>253</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>54</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.016006</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>66</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.010083</td>\n",
       "      <td>1228</td>\n",
       "      <td>409</td>\n",
       "      <td>Q</td>\n",
       "      <td>L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>78</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.016263</td>\n",
       "      <td>1539</td>\n",
       "      <td>513</td>\n",
       "      <td>N</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>88</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.013609</td>\n",
       "      <td>864</td>\n",
       "      <td>288</td>\n",
       "      <td>P</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>89</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.027779</td>\n",
       "      <td>1335</td>\n",
       "      <td>445</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>89</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.010439</td>\n",
       "      <td>869</td>\n",
       "      <td>289</td>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>103</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.056567</td>\n",
       "      <td>1120</td>\n",
       "      <td>373</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Id ref var  freq.var   pos  aa_pos aa_consensus aa_var\n",
       "34    3   A   T  0.020529  1114     371            Q      L\n",
       "7     3   G   G  0.014898   670     223            K      R\n",
       "2     6   C   A  0.039535   284      94            F      L\n",
       "13   10   A   G  0.034423   687     229            I      V\n",
       "29   10   G   A  0.019396  1014     338            V      I\n",
       "32   12   G   T  0.017493  1047     349            A      S\n",
       "21   14   G   A  0.078873   853     284            R      K\n",
       "68   17   A   A  0.422878  1613     537            M      I\n",
       "0    27   G   A  0.030421   192      64            G      S\n",
       "28   46   A   G  0.030150   879     293            N      D\n",
       "20   50   A   C  0.028811   761     253            K      N\n",
       "9    54   G   G  0.016006   670     223            K      R\n",
       "43   66   A   T  0.010083  1228     409            Q      L\n",
       "62   78   A   C  0.016263  1539     513            N      H\n",
       "24   88   C   T  0.013609   864     288            P      S\n",
       "50   89   G   A  0.027779  1335     445            A      T\n",
       "25   89   T   G  0.010439   869     289            I      M\n",
       "35  103   C   A  0.056567  1120     373            S      Y"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nonsense=df.loc[(df.aa_consensus!=df.aa_var)]\n",
    "nonsense[[\"Id\",\"ref\",'var',\"freq.var\",\"pos\",\"aa_pos\",\"aa_consensus\",\"aa_var\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## H3 numbering system! "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here I am adpapting the HA number script made by Jesse Bloom of the Bloom lab. Currently his script runs as an exicutable that takes in a file containing variables - sequence, position of interest, and path to aligner. I will adapt this into a function to be used here. I will also have to alter the output. [Currently I am just grabbing some of the text output as formating in his script and saving it in lists to be added to the data frame.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import tempfile\n",
    "import sys\n",
    "import subprocess"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adapted HA_numbering.py - function "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def Align(headers_seqs, progpath, program='PROBCONS', musclegapopen=None):\n",
    "    \"\"\"Performs a multiple sequence alignment of two or more sequences.\n",
    "\n",
    "    By default, the protein sequences are aligned using PROBCONS.  This is\n",
    "        probably the most accurate alignment program.  However, it is\n",
    "        slow and consumes large amounts of memory if you are aligning\n",
    "        a very large number of sequences (typically if you are aligning\n",
    "        more than several hundred).  In that case, you may prefer to use\n",
    "        MUSCLE instead.  You can choose between the two with the 'program'\n",
    "        option.  If you decide to use MUSCLE, you can also align nucleotide\n",
    "        sequences with this program.\n",
    "\n",
    "    'headers_seqs' is a list specifying the names of the sequences that we\n",
    "        want to align.  Each entry is a 2-tuple '(head, seq)' where 'head' is\n",
    "        a header giving the sequence name and other information (might be empty)\n",
    "        and 'seq' is a string giving the protein sequence.  The list must have\n",
    "        at least 2 entries.\n",
    "\n",
    "    'progpath' should specify a directory containing the alignment program executable,\n",
    "        either PROBCONS or MUSCLE.  The PROBCONS executable is assumed to have\n",
    "        the name \"probcons\" in this directory.  The MUSCLE executable is assumed to\n",
    "        have the name \"muscle\" in this directory.\n",
    "\n",
    "    'program' specifies what program to use for the alignment.  By default, it is\n",
    "        \"PROBCONS\".  If you wish to use MUSCLE instead, set it to \"MUSCLE\".  \n",
    "\n",
    "    'musclegapopen' sets the MUSCLE gap openining penalty to the specified\n",
    "        value. By default it is None, meaning we use the MUSCLE default penalty.\n",
    "        You can also set it to a number; for example -100 will lead to fewer gaps.\n",
    "\n",
    "    This executable is used to perform a multiple sequence alignment of the proteins\n",
    "        with the default settings of either PROBCONS or MUSCLE.  The returned variable is a\n",
    "        new list 'aligned_headers_seqs'.  Each entry is a 2-tuple '(head, aligned_seq)'.  \n",
    "        'head' has the same meaning as on input (the sequence header) and \n",
    "        'aligned_seq' is the aligned sequence, with gaps inserted as '-'\n",
    "        as appropriate.  Therefore, all of the 'aligned_seq' entries in\n",
    "        'aligned_headers_seqs' are of the same length.  The entries in 'aligned_headers_seq'\n",
    "        are in the same order as in the input list 'headers_seqs'.\n",
    "    \"\"\"\n",
    "    if not (isinstance(headers_seqs, list) and len(headers_seqs) >= 2):\n",
    "        raise ValueError('header_seqs does not specify a list with at least two entries.')\n",
    "    if not os.path.isdir(progpath):\n",
    "        raise ValueError(\"Cannot find directory %s.\" % progpath)\n",
    "    if program == 'PROBCONS':\n",
    "        exe = os.path.abspath(\"%s/probcons\" % progpath) # the executable\n",
    "    elif program == 'MUSCLE':\n",
    "        exe = os.path.abspath(\"%s/muscle\" % progpath) # the executable\n",
    "    else:\n",
    "        raise ValueError(\"Invalid value of %s for 'program'.\" % (str(program)))\n",
    "    if not os.path.isfile(exe):\n",
    "        raise IOError(\"Cannot find executable at %s.\" % exe)\n",
    "    currdir = os.getcwd()\n",
    "    tempdir = tempfile.mkdtemp()\n",
    "    try:\n",
    "        # do stuff in a temporary directory\n",
    "        infile = \"%s/in.fasta\" % tempdir # input file \n",
    "        outfile = \"%s/out.fasta\" % tempdir # output file \n",
    "        WriteFASTA(headers_seqs, infile) # write sequences to the input file\n",
    "        if program == 'PROBCONS':\n",
    "            p = subprocess.Popen(\"%s %s\" % (exe, infile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run ProbCons\n",
    "            (output, errors) = p.communicate()\n",
    "            \n",
    "            open(outfile, 'wb').write(output)\n",
    "        elif program == 'MUSCLE':\n",
    "            if musclegapopen != None:\n",
    "                p = subprocess.Popen(\"%s -gapopen %d -in %s -out %s\" % (exe, musclegapopen, infile, outfile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run MUSCLE\n",
    "            else:\n",
    "                p = subprocess.Popen(\"%s -in %s -out %s\" % (exe, infile, outfile), shell = True, stdout = subprocess.PIPE, stderr = subprocess.PIPE) # run MUSCLE\n",
    "            (output, errors) = p.communicate()\n",
    "        try:\n",
    "            aligned_headers_seqs = ReadFASTA(outfile)\n",
    "        except:\n",
    "            sys.stderr.write(\"Error getting alignment output, error of %s\" % errors)\n",
    "            raise\n",
    "    finally:\n",
    "        os.chdir(currdir) # return to the original directory\n",
    "        for file in os.listdir(tempdir):\n",
    "            os.remove(\"%s/%s\" % (tempdir, file)) # remove files from temporary directory\n",
    "        os.rmdir(tempdir) # remove temporary directory\n",
    "    if len(aligned_headers_seqs) != len(headers_seqs):\n",
    "        raise ValueError(\"Did not return the correct number of aligned sequences.\")\n",
    "    # put the aligned sequences in the same order as the input sequences\n",
    "    n = len(aligned_headers_seqs[0][1]) # length of aligned sequences\n",
    "    d = dict(aligned_headers_seqs)\n",
    "    aligned_headers_seqs = []\n",
    "    for (head, seq) in headers_seqs:\n",
    "        try:\n",
    "            alignedseq = d[head]\n",
    "        except KeyError:\n",
    "            raise ValueError(\"After alignment, the following header is missing: %s\" % head)\n",
    "        if len(alignedseq) != n:\n",
    "            open('errors.temp', 'w').write(errors)\n",
    "            raise ValueError(\"Aligned sequence %s is not of length %d: if you are using MUSCLE, you may be running out of memory.  Errors have been written to errors.temp.\" % (alignedseq, n))\n",
    "        if len(seq) > n:\n",
    "            open('errors.temp', 'w').write(errors)\n",
    "            raise ValueError(\"Unaligned seq %s is longer than aligned length of %d: if you are using MUSCLE, you many be running out of memory.  Errors have been written to errors.temp.\" % (seq, n))\n",
    "        aligned_headers_seqs.append((head, alignedseq))\n",
    "    return aligned_headers_seqs # return the aligned sequences\n",
    "    \n",
    "\n",
    "def StripGapsToFirstSequence(aligned_headers_seqs):\n",
    "    \"\"\"Strips gaps from a reference sequence, and all corresponding alignments.\n",
    "\n",
    "    On input, 'aligned_headers_seqs' should be a set of two or more aligned sequences,\n",
    "        as would be returned by 'Align'.\n",
    "\n",
    "    The first sequence in this alignment is taken to correspond to the reference sequence.\n",
    "        The returned variable is a list similar to aligned_headers_seqs, but with\n",
    "        all positions corresponding to gaps in this reference sequence stripped away.\n",
    "        All gaps ('-') characters are removed from this reference sequence.  In addition,\n",
    "        in all other aligned sequences in 'aligned_headers_seqs', every character at\n",
    "        the same position as a gap in the reference sequence is removed.  Therefore,\n",
    "        at the end of this procedure, all of the alignments have the same length\n",
    "        as the reference sequence with its gaps stripped away.  The headers are \n",
    "        unchanged.  The order of sequences in this stripped alignment is also\n",
    "        unchanged.\n",
    "\n",
    "    >>> StripGapsToFirstSequence([('s1', '-AT-A-GC'), ('s2', 'AAT-TAGC'), ('s3', '--T-A-GC')])\n",
    "    [('s1', 'ATAGC'), ('s2', 'ATTGC'), ('s3', '-TAGC')]\n",
    "    \"\"\"        \n",
    "    if not (isinstance(aligned_headers_seqs, list) and len(aligned_headers_seqs) >= 2):\n",
    "        raise ValueError(\"aligned_headers_seqs does not specify at least two aligned sequences.\")\n",
    "    (ref_head, ref_seq) = aligned_headers_seqs[0]\n",
    "    non_strip_positions = [] # positions not to strip away\n",
    "    stripped_ref_seq = []\n",
    "    for i in range(len(ref_seq)):\n",
    "        r = ref_seq[i]\n",
    "        if r != '-':\n",
    "            non_strip_positions.append(i)\n",
    "            stripped_ref_seq.append(r)\n",
    "    stripped_headers_seqs = [(ref_head, ''.join(stripped_ref_seq))]\n",
    "    for (ihead, iseq) in aligned_headers_seqs[1 : ]:\n",
    "        istrippedseq = ''.join([iseq[i] for i in non_strip_positions])\n",
    "        stripped_headers_seqs.append((ihead, istrippedseq))\n",
    "    return stripped_headers_seqs\n",
    "\n",
    "\n",
    "def WriteFASTA(headers_seqs, filename, writable_file=False):\n",
    "    \"\"\"Writes sequences to a FASTA file.\n",
    "\n",
    "    'headers_seqs' is a list of 2-tuples specifying sequences and their\n",
    "        corresponding headers.  Each entry is the 2-tuple '(header, seq)'\n",
    "        where 'header' is a string giving the header (without the leading \">\"),\n",
    "        and 'seq' is the corresponding sequence.\n",
    "\n",
    "    'filename' is a string that specifies the name of the file to which the\n",
    "        headers and sequences should be written.  If this file already exists,\n",
    "        it is overwritten. \n",
    "\n",
    "    'writable_file' is a Boolean switch specifying that rather than 'filename'\n",
    "        giving a string specifying the name of a file to which the sequences\n",
    "        should be written, it instead specifies a writable file object to which\n",
    "        the sequences should be written.\n",
    "\n",
    "    The sequences are written to the file in the same order that they are specified\n",
    "        in 'headers_seqs'.\n",
    "    \"\"\"\n",
    "    assert isinstance(writable_file, bool)\n",
    "    if writable_file:\n",
    "        f = filename\n",
    "    else:\n",
    "        f = open(filename, 'w')\n",
    "    for (header, seq) in headers_seqs:\n",
    "        f.write(\">%s\\n%s\\n\" % (header, seq))\n",
    "    if not writable_file:\n",
    "        f.close()\n",
    "\n",
    "\n",
    "\n",
    "def ReadFASTA(fastafile):\n",
    "    \"\"\"Reads sequences from a FASTA file.\n",
    "\n",
    "    'fastafile' should specify the name of a FASTA file.\n",
    "\n",
    "    This function reads all sequences from the FASTA file.  It returns the\n",
    "        list 'headers_seqs'.  This list is composed of a 2-tuple '(header, seq)'\n",
    "        for every sequence entry in FASTA file.  'header' is the header for\n",
    "        a sequence, with the leading \">\" and any trailing spaces removed. 'seq'\n",
    "        is the corresponding sequence.\n",
    "    \"\"\"\n",
    "    lines = open(fastafile).readlines()\n",
    "    headers_seqs = []\n",
    "    header = None\n",
    "    seq = []\n",
    "    for line in lines:\n",
    "        if line[0] == '>':\n",
    "            if (not header) and (not seq):\n",
    "                pass # first sequence in file\n",
    "            elif header and not seq:\n",
    "                raise ValueError(\"Empty sequence for %s\" % header)\n",
    "            elif seq and not header:\n",
    "                raise ValueError(\"File does not begin with header.\")\n",
    "            else:\n",
    "                seq = ''.join(seq)\n",
    "                seq = seq.replace(' ', '')\n",
    "                headers_seqs.append((header, seq))\n",
    "            header = line.strip()[1 : ]\n",
    "            seq = []\n",
    "        else:\n",
    "            seq.append(line.strip())\n",
    "    if (not header) and (not seq):\n",
    "        pass # first sequence in file\n",
    "    elif header and not seq:\n",
    "        raise ValueError(\"Empty sequence for %s\" % header)\n",
    "    elif seq and not header:\n",
    "        raise ValueError(\"File does not begin with header.\")\n",
    "    else:\n",
    "        seq = ''.join(seq)\n",
    "        seq = seq.replace(' ', '')\n",
    "        headers_seqs.append((header, seq))\n",
    "    return headers_seqs\n",
    "\n",
    "\n",
    "def GetCorrespondingResidue(seqs, i):\n",
    "    \"\"\"Gets the corresponding residue number for two aligned sequences.\n",
    "\n",
    "    *seqs* is a set of two aligned sequences as *(head, sequence)* 2-tuples.\n",
    "\n",
    "    *i* is the number of a residue in sequential numbering of *seqs[0]*\n",
    "    without considering any of the gaps induced by alignment, in 1, 2, ...\n",
    "    numbering.\n",
    "\n",
    "    Returns the number of the residue in sequential numbering of *seqs[1]*\n",
    "    without considering any of the gaps induced by alignment in 1, 2, ...\n",
    "    numbering. Returns *None* if residue *i* of *seqs[0]* aligns with a\n",
    "    gap in *seqs[1]*.\n",
    "    \"\"\"\n",
    "    assert len(seqs) == 2\n",
    "    s1 = seqs[0][1]\n",
    "    s2 = seqs[1][1]\n",
    "    assert len(s1) == len(s2)\n",
    "    assert 1 <= i <= len(s1)\n",
    "    s1index = s2index = 0\n",
    "    for j in range(len(s1)):\n",
    "        if s1[j] != '-':\n",
    "            s1index += 1\n",
    "        if s2[j] != '-':\n",
    "            s2index += 1\n",
    "        if s1index == i:\n",
    "            if s2[j] == '-':\n",
    "                return None\n",
    "            else:\n",
    "                return s2index\n",
    "\n",
    "\n",
    "def find_aligments(aligner,aligner_path,seq,pos): # The positions will be given as base 0 and adjusted to match the convention (base 1) in the funciton\n",
    "    \"\"\"Main body of script.\"\"\"\n",
    "    print(\"\\nBeginning execution of HA_numbering script.\")\n",
    "    \n",
    "    # parse arguments\n",
    "    if aligner == 'probcons':\n",
    "        alignerpath = aligner_path # redundant\n",
    "        if not os.path.isdir(alignerpath):\n",
    "            raise IOError(\"The directory of %s specified by probconspath does not exist.\" % (alignerpath))\n",
    "        prog = 'PROBCONS'\n",
    "    elif aligner == 'muscle':\n",
    "        alignerpath = aligner_path # redundant\n",
    "        if not os.path.isdir(alignerpath):\n",
    "            raise IOError(\"The directory of %s specified by musclepath does not exist.\" % (alignerpath))\n",
    "        prog = 'MUSCLE'\n",
    "    else:\n",
    "        raise IOError(\"First argument does not specify probcons or muscle\")\n",
    "    ha_sequence = str(seq) # just get the sequence\n",
    "    sites = pos\n",
    "    # Define sequences and their numbering conversions.\n",
    "    # The sequences are in seq_d and keyed by PDB code.\n",
    "    # The number conversions from sequential numbering of these sequences to\n",
    "    # th number labels are in label_d and keyed by PDB codes.\n",
    "    seq_names = ['4HMG', '4JTV']\n",
    "    seq_d = {}\n",
    "    label_d = {}\n",
    "    seq_4hmg_a = \\\n",
    "        'QDLPGNDNSTATLCLGHHAVPNGTLVKTITDDQIEVTNATELVQSSSTGKICNNPHRILDGIDCTLIDALLGDPHCDVFQ' +\\\n",
    "        'NETWDLFVERSKAFSNCYPYDVPDYASLRSLVASSGTLEFITEGFTWTGVTQNGGSNACKRGPGSGFFSRLNWLTKSGST' +\\\n",
    "        'YPVLNVTMPNNDNFDKLYIWGIHHPSTNQEQTSLYVQASGRVTVSTRRSQQTIIPNIGSRPWVRGQSSRISIYWTIVKPG' +\\\n",
    "        'DVLVINSNGNLIAPRGYFKMRTGKSSIMRSDAPIDTCISECITPNGSIPNDKPFQNVNKITYGACPKYVKQNTLKLATGM' +\\\n",
    "        'RNVPEKQT'\n",
    "    seq_4hmg_b = \\\n",
    "        'GLFGAIAGFIENGWEGMIDGWYGFRHQNSEGTGQAADLKSTQAAIDQINGKLNRVIEKTNEKFHQIEKEFSEVEGRIQDL' +\\\n",
    "        'EKYVEDTKIDLWSYNAELLVALENQHTIDLTDSEMNKLFEKTRRQLRENAEEMGNGCFKIYHKCDNACIESIRNGTYDHD' +\\\n",
    "        'VYRDEALNNRFQIKG'\n",
    "    seq_d['4HMG'] = seq_4hmg_a + seq_4hmg_b\n",
    "    label_d['4HMG'] = dict([(i + 1, '%d in HA1' % (i + 1)) for i in range(len(seq_4hmg_a))] + [(len(seq_4hmg_a) + i + 1, '%d in HA2' % (i + 1)) for i in range(len(seq_4hmg_b))])\n",
    "    assert len(seq_d['4HMG']) == len(label_d['4HMG'])\n",
    "    seq_4jtv_a = \\\n",
    "        'DTLCIGYHANNSTDTVDTVLEKNVTVTHSVNLLEDKHNGKLCKLRGVAPLHLGKCNIAGWILGNPECESLSTASSWSYIV' +\\\n",
    "        'ETPSSDNGTCYPGDFIDYEELREQLSSVSSFERFEIFPKTSSWPNHDSNKGVTAACPHAGAKSFYKNLIWLVKKGNSYPK' +\\\n",
    "        'LSKSYINDKGKEVLVLWGIHHPSTSADQQSLYQNADTYVFVGSSRYSKKFKPEIAIRPKVRDQEGRMNYYWTLVEPGDKI' +\\\n",
    "        'TFEATGNLVVPRYAFAMERNAGSGIIISDTPVHDCNTTCQTPKGAINTSLPFQNIHPITIGKCPKYVKSTKLRLATGLRN' +\\\n",
    "        'I'\n",
    "    seq_4jtv_b = \\\n",
    "        'GLFGAIAGFIEGGWTGMVDGWYGYHHQNEQGSGYAADLKSTQNAIDEITNKVNSVIEKMNTQFTAVGKEFNHLEKRIENL' +\\\n",
    "        'NKKVDDGFLDIWTYNAELLVLLENERTLDYHDSNVKNLYEKVRSQLKNNAKEIGNGCFEFYHKCDNTCMESVKNGTYDYP' +\\\n",
    "        'KY'\n",
    "    seq_d['4JTV'] = seq_4jtv_a + seq_4jtv_b\n",
    "    label_d['4JTV'] = dict([(i + 1, '%d of HA1' % (i + 7)) for i in range(len(seq_4jtv_a))] + [(len(seq_4jtv_a) + i + 1, '%d of HA2' % (i + 1)) for i in range(len(seq_4jtv_b))])\n",
    "    assert len(seq_d['4JTV']) == len(label_d['4JTV'])\n",
    "\n",
    "    # make alignments\n",
    "    print(\"Making %s alignments...\" % prog)\n",
    "    alignments = {}\n",
    "    for seqname in seq_names:\n",
    "        alignments[seqname] = Align([('seq', ha_sequence), (seqname, seq_d[seqname])], alignerpath, prog)\n",
    "    print(\"Alignments complete.\\n\\nHere are the corresponding residue numbers:\")\n",
    "    for site in sites:\n",
    "        site=site+1 # to correct for difference in base numbering\n",
    "        if not (1 <= site <= len(ha_sequence)):\n",
    "            raise ValueError(\"site %d is outside the valid range for sequential numbering of ha_sequence starting at 1.\" % site)\n",
    "        sitestring = ['\\nResidue %s%d in sequential numbering of your HA sequence corresponds to:' % (ha_sequence[site - 1], site)]\n",
    "        for seqname in seq_names:\n",
    "            i = GetCorrespondingResidue(alignments[seqname], site)\n",
    "            if i == None:\n",
    "                sitestring.append('  * an alignment gap in %s' % seqname) \n",
    "            else:\n",
    "                if not (1 <= i <= len(seq_d[seqname])):\n",
    "                    raise ValueError(\"Invalid corrresponding residue for %s -- something is wrong with this program\" % seqname)\n",
    "                aa = seq_d[seqname][i - 1]\n",
    "                \n",
    "                sitestring.append('  * %s%s in %s' % (aa, label_d[seqname][i], seqname))\n",
    "        print( '\\n'.join(sitestring))\n",
    "        return([sitestring[1][4:-7],sitestring[2][4:-7]])\n",
    "        \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now I am looping through all nonsynonymous mutations as before and applying the function above to get the H3 numbering system. The script takes in base 0 positions and handels them accordingly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L372 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q27 in HA2 in 4HMG\n",
      "  * Q27 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K224 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R208 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L95 in sequential numbering of your HA sequence corresponds to:\n",
      "  * F79 in HA1 in 4HMG\n",
      "  * L76 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V230 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I214 in HA1 in 4HMG\n",
      "  * K217 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue V339 in sequential numbering of your HA sequence corresponds to:\n",
      "  * V323 in HA1 in 4HMG\n",
      "  * I327 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S350 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A5 in HA2 in 4HMG\n",
      "  * A5 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue K285 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R269 in HA1 in 4HMG\n",
      "  * I273 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I538 in sequential numbering of your HA sequence corresponds to:\n",
      "  * an alignment gap in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S65 in sequential numbering of your HA sequence corresponds to:\n",
      "  * G49 in HA1 in 4HMG\n",
      "  * G45 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue D294 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I278 in HA1 in 4HMG\n",
      "  * N282 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue N254 in sequential numbering of your HA sequence corresponds to:\n",
      "  * K238 in HA1 in 4HMG\n",
      "  * E241 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue R224 in sequential numbering of your HA sequence corresponds to:\n",
      "  * R208 in HA1 in 4HMG\n",
      "  * R211 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue L410 in sequential numbering of your HA sequence corresponds to:\n",
      "  * Q65 in HA2 in 4HMG\n",
      "  * A65 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue H514 in sequential numbering of your HA sequence corresponds to:\n",
      "  * N169 in HA2 in 4HMG\n",
      "  * an alignment gap in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue S289 in sequential numbering of your HA sequence corresponds to:\n",
      "  * P273 in HA1 in 4HMG\n",
      "  * P277 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue T446 in sequential numbering of your HA sequence corresponds to:\n",
      "  * A101 in HA2 in 4HMG\n",
      "  * L101 of HA2 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue I290 in sequential numbering of your HA sequence corresponds to:\n",
      "  * I274 in HA1 in 4HMG\n",
      "  * V278 of HA1 in 4JTV\n",
      "\n",
      "Beginning execution of HA_numbering script.\n",
      "Making PROBCONS alignments...\n",
      "Alignments complete.\n",
      "\n",
      "Here are the corresponding residue numbers:\n",
      "\n",
      "Residue Y374 in sequential numbering of your HA sequence corresponds to:\n",
      "  * S29 in HA2 in 4HMG\n",
      "  * E29 of HA2 in 4JTV\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.5/site-packages/ipykernel/__main__.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/usr/local/lib/python3.5/site-packages/ipykernel/__main__.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "\n",
    "counter=0\n",
    "PDB_4HMG=[]\n",
    "PDB_4JTV=[]\n",
    "for id in nonsense.Id:\n",
    "    if str(id) in dn_Ids:\n",
    "        i=dn_Ids.index(str(id)) # get the index of the sample\n",
    "        pos=list(nonsense[\"aa_pos\"])[counter]\n",
    "        nice=find_aligments('probcons','/Users/jt/probcons/',dn_protein[i],[pos])\n",
    "        PDB_4HMG.append(nice[0])\n",
    "        PDB_4JTV.append(nice[1])\n",
    "    counter=counter+1        \n",
    "nonsense[\"PDB_4HMG\"] = PDB_4HMG\n",
    "nonsense[\"PDB_4JTV\"]=PDB_4JTV\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Final output!! "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Id</th>\n",
       "      <th>ref</th>\n",
       "      <th>var</th>\n",
       "      <th>freq.var</th>\n",
       "      <th>pos</th>\n",
       "      <th>aa_pos</th>\n",
       "      <th>aa_consensus</th>\n",
       "      <th>aa_var</th>\n",
       "      <th>PDB_4HMG</th>\n",
       "      <th>PDB_4JTV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.020529</td>\n",
       "      <td>1114</td>\n",
       "      <td>371</td>\n",
       "      <td>Q</td>\n",
       "      <td>L</td>\n",
       "      <td>Q27 in HA2</td>\n",
       "      <td>Q27 of HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>3</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.014898</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>R211 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.039535</td>\n",
       "      <td>284</td>\n",
       "      <td>94</td>\n",
       "      <td>F</td>\n",
       "      <td>L</td>\n",
       "      <td>F79 in HA1</td>\n",
       "      <td>L76 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.034423</td>\n",
       "      <td>687</td>\n",
       "      <td>229</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>I214 in HA1</td>\n",
       "      <td>K217 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.019396</td>\n",
       "      <td>1014</td>\n",
       "      <td>338</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>V323 in HA1</td>\n",
       "      <td>I327 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>12</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.017493</td>\n",
       "      <td>1047</td>\n",
       "      <td>349</td>\n",
       "      <td>A</td>\n",
       "      <td>S</td>\n",
       "      <td>A5 in HA2</td>\n",
       "      <td>A5 of HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>14</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.078873</td>\n",
       "      <td>853</td>\n",
       "      <td>284</td>\n",
       "      <td>R</td>\n",
       "      <td>K</td>\n",
       "      <td>R269 in HA1</td>\n",
       "      <td>I273 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>17</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>0.422878</td>\n",
       "      <td>1613</td>\n",
       "      <td>537</td>\n",
       "      <td>M</td>\n",
       "      <td>I</td>\n",
       "      <td>an alignment gap</td>\n",
       "      <td>an alignment gap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.030421</td>\n",
       "      <td>192</td>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>G49 in HA1</td>\n",
       "      <td>G45 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>46</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.030150</td>\n",
       "      <td>879</td>\n",
       "      <td>293</td>\n",
       "      <td>N</td>\n",
       "      <td>D</td>\n",
       "      <td>I278 in HA1</td>\n",
       "      <td>N282 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>50</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.028811</td>\n",
       "      <td>761</td>\n",
       "      <td>253</td>\n",
       "      <td>K</td>\n",
       "      <td>N</td>\n",
       "      <td>K238 in HA1</td>\n",
       "      <td>E241 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>54</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>0.016006</td>\n",
       "      <td>670</td>\n",
       "      <td>223</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>R208 in HA1</td>\n",
       "      <td>R211 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>66</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.010083</td>\n",
       "      <td>1228</td>\n",
       "      <td>409</td>\n",
       "      <td>Q</td>\n",
       "      <td>L</td>\n",
       "      <td>Q65 in HA2</td>\n",
       "      <td>A65 of HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>78</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.016263</td>\n",
       "      <td>1539</td>\n",
       "      <td>513</td>\n",
       "      <td>N</td>\n",
       "      <td>H</td>\n",
       "      <td>N169 in HA2</td>\n",
       "      <td>an alignment gap</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>88</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.013609</td>\n",
       "      <td>864</td>\n",
       "      <td>288</td>\n",
       "      <td>P</td>\n",
       "      <td>S</td>\n",
       "      <td>P273 in HA1</td>\n",
       "      <td>P277 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>89</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.027779</td>\n",
       "      <td>1335</td>\n",
       "      <td>445</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>A101 in HA2</td>\n",
       "      <td>L101 of HA2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>89</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>0.010439</td>\n",
       "      <td>869</td>\n",
       "      <td>289</td>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "      <td>I274 in HA1</td>\n",
       "      <td>V278 of HA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>103</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.056567</td>\n",
       "      <td>1120</td>\n",
       "      <td>373</td>\n",
       "      <td>S</td>\n",
       "      <td>Y</td>\n",
       "      <td>S29 in HA2</td>\n",
       "      <td>E29 of HA2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Id ref var  freq.var   pos  aa_pos aa_consensus aa_var  \\\n",
       "34    3   A   T  0.020529  1114     371            Q      L   \n",
       "7     3   G   G  0.014898   670     223            K      R   \n",
       "2     6   C   A  0.039535   284      94            F      L   \n",
       "13   10   A   G  0.034423   687     229            I      V   \n",
       "29   10   G   A  0.019396  1014     338            V      I   \n",
       "32   12   G   T  0.017493  1047     349            A      S   \n",
       "21   14   G   A  0.078873   853     284            R      K   \n",
       "68   17   A   A  0.422878  1613     537            M      I   \n",
       "0    27   G   A  0.030421   192      64            G      S   \n",
       "28   46   A   G  0.030150   879     293            N      D   \n",
       "20   50   A   C  0.028811   761     253            K      N   \n",
       "9    54   G   G  0.016006   670     223            K      R   \n",
       "43   66   A   T  0.010083  1228     409            Q      L   \n",
       "62   78   A   C  0.016263  1539     513            N      H   \n",
       "24   88   C   T  0.013609   864     288            P      S   \n",
       "50   89   G   A  0.027779  1335     445            A      T   \n",
       "25   89   T   G  0.010439   869     289            I      M   \n",
       "35  103   C   A  0.056567  1120     373            S      Y   \n",
       "\n",
       "             PDB_4HMG           PDB_4JTV  \n",
       "34        Q27 in HA2         Q27 of HA2   \n",
       "7        R208 in HA1        R211 of HA1   \n",
       "2         F79 in HA1         L76 of HA1   \n",
       "13       I214 in HA1        K217 of HA1   \n",
       "29       V323 in HA1        I327 of HA1   \n",
       "32         A5 in HA2          A5 of HA2   \n",
       "21       R269 in HA1        I273 of HA1   \n",
       "68  an alignment gap   an alignment gap   \n",
       "0         G49 in HA1         G45 of HA1   \n",
       "28       I278 in HA1        N282 of HA1   \n",
       "20       K238 in HA1        E241 of HA1   \n",
       "9        R208 in HA1        R211 of HA1   \n",
       "43        Q65 in HA2         A65 of HA2   \n",
       "62       N169 in HA2   an alignment gap   \n",
       "24       P273 in HA1        P277 of HA1   \n",
       "50       A101 in HA2        L101 of HA2   \n",
       "25       I274 in HA1        V278 of HA1   \n",
       "35        S29 in HA2         E29 of HA2   "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(list(nonsense[\"Id\"])))\n",
    "nonsense[[\"Id\",\"ref\",'var',\"freq.var\",\"pos\",\"aa_pos\",\"aa_consensus\",\"aa_var\",\"PDB_4HMG\",\"PDB_4JTV\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The final step will be to use this list to reference the putative antigenic sites to see if any match.\n",
    "\n",
    "\n",
    "\n",
    "And this saves the output as a csv."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "nonsense[[\"Id\",\"ref\",'var',\"freq.var\",\"pos\",\"aa_pos\",\"aa_consensus\",\"aa_var\",\"PDB_4HMG\",\"PDB_4JTV\"]].to_csv(\"../data/nonsymom.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
